woocommerce/woocommerce-android
# Media picker library integration

This is a first take on integration of the new [Media picker library](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android). Visually, the experience should be very similar, but the entire media-selection workflow is now governed by the library, including permission request handling.

A few things to note: 

- All logs from the MP library are delegated through the [`Log`](WooCommerce/src/main/kotlin/com/woocommerce/android/tools/MediaPickerLogger.kt) interface to `AppLog`.
- All tracks from the library are delegated through the [`Tracker`](WooCommerce/src/main/kotlin/com/woocommerce/android/tools/MediaPickerTracker.kt) interface to `AnalyticsTracker`.
- Currently only the local media and camera are handled by the library now. The WP media library is still using the old code because the WP media source hasn't been implemented yet and it will be replaced next.
- **detekt** kept complaining about the complexity for the [enum mapper](https://github.com/woocommerce/woocommerce-android/pull/5033/files#diff-cd0f3150d17ea6f8d970eb48a63780df8926e809d1f6f23c0c8852f2c8b71cbcR13), so I added an exception because there isn't really anything to fix.
- This is a draft because it's a new library and I wanted to gather some feedback before we decided to use it in production.

**To test:**

- [ ] _Permissions_

1. Disable both STORAGE and CAMERA permissions in the settings and try adding local media
2. Allow the requests and notice the local media is displayed
3. Disable the permissions and deny the requests repeatedly
4. Notice a dialog is displayed
5. Go back and try it again
6. Notice a button to open the app settings is shown

- [ ] _Image selection_
1. With the permissions enabled now, try adding a local media
2. Notice the images are loaded
3. Tap on the System picker icon and notice it is loaded
4. Try selecting some images
5. Notice the images are selected and uploaded
6. Add more images from the local media picker and notice they're uploaded
7. Try adding images using the camera and notice the

- [ ] _Search_
1. Select the local device media option
2. Tap on the search icon
3. Type a string matching a subset of images
4. Notice the images are filtered based on the string

- [ ] _Tracks_

Verify all tracks events are logged:

-  MEDIA_PICKER_PREVIEW_OPENED
-  MEDIA_PICKER_RECENT_MEDIA_SELECTED
-  MEDIA_PICKER_OPEN_GIF_LIBRARY
-  MEDIA_PICKER_OPEN_DEVICE_LIBRARY
-  MEDIA_PICKER_CAPTURE_PHOTO
-  MEDIA_PICKER_SEARCH_TRIGGERED
-  MEDIA_PICKER_SEARCH_EXPANDED
-  MEDIA_PICKER_SEARCH_COLLAPSED
-  MEDIA_PICKER_SHOW_PERMISSIONS_SCREEN
-  MEDIA_PICKER_ITEM_SELECTED
-  MEDIA_PICKER_ITEM_UNSELECTED
-  MEDIA_PICKER_SELECTION_CLEARED
-  MEDIA_PICKER_OPENED
-  MEDIA_PICKER_OPEN_SYSTEM_PICKER
